job_creator.php

```diff
@@ -52,59 +56,47 @@ public function getInstallerVersion(
-                $composerRepo = 'silverstripe/' . str_replace('silverstripe-', '', $lockedSteppedRepo);
-                if (isset($json->require->{$composerRepo})) {
-                    $version = $json->require->{$composerRepo};
-                    if (preg_match('#^([0-9\.]+)\.x\-dev$#', $version, $matches)) {
-                        $versionNumber = $matches[1];
-                        // If the lockstepped dependency is "three less" (e.g. silverstripe/admin is 3 major
-                        // versions behind silverstripe/installer), account for that here.
-                        if (in_array($lockedSteppedRepo, LOCKSTEPPED_REPOS_VERSION_IS_THREE_LESS)) {
-                            $versionNumber += 3;
+            foreach ($this->lockSteppedRepos as $lockedSteppedRepo => $majorVersionMapping) {
+                if (isset($json->require->{$lockedSteppedRepo})) {
+                    $version = $json->require->{$lockedSteppedRepo};
+                    if (preg_match('#^([0-9]+)(\.[0-9]+)?\.x\-dev$#', $version, $matches)) {
+                        $dependencyMajorVersion = $matches[1];
+                        $dependencyMinorSuffix = $matches[2] ?? null;
+                        $versionNumber = null;
+                        foreach ($majorVersionMapping as $cmsMajorFromMap => $repoBranches) {
+                            if (is_numeric($cmsMajorFromMap) && in_array($dependencyMajorVersion, $repoBranches)) {
+                                $versionNumber = $cmsMajorFromMap . $dependencyMinorSuffix;
+                                break;
+                            }
+                        }
+                        if ($versionNumber !== null) {
+                            return $versionNumber . '.x-dev';
                         }
-                        return $versionNumber . '.x-dev';
                     }
                 }
             }
         }
         // fallback to use the next-minor or latest-minor version of installer
-        $installerVersions = array_keys(INSTALLER_TO_PHP_VERSIONS);
+        $installerVersions = array_keys(MetaData::PHP_VERSIONS_FOR_CMS_RELEASES);
         $installerVersions = array_filter($installerVersions, fn($version) => substr($version, 0, 1) === $cmsMajor);
 
-        if (preg_match('#^[1-9]+[0-9]*$#', $branch)) {
+        if (preg_match('#^[1-9]+[0-9]*$#', $this->branch)) {
             // next-minor e.g. 4
             return $cmsMajor . '.x-dev';
         } else {
             // current-minor e.g. 4.11
             // remove major versions
-            $installerVersions = array_diff($installerVersions, ['4', '5', '6', '7', '8', '9']);
-            // get the minor portions of the verisons e.g. [9, 10, 11]
+            $installerVersions = array_filter($installerVersions, fn ($v) => !ctype_digit((string) $v));
+            // get the minor portions of the versions e.g for ['4.9', '4.10', '4.11'] this returns [9, 10, 11]
```

Comment:
Unrelated refactoring that makes this more robust - we don't care about those specific values, we just care whether the value is a major (all digits) or not.